---
title: 'motifbreakR: an Introduction'
author: "Simon G. Coetzee, Gerhard A. Coetzee and Dennis J. Hazelett"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: true
    highlight: textmate
    keep_md: yes
    theme: flatly
    toc: yes
    fig_width: 4.5
    fig_height: 4
  BiocStyle::pdf_document:
    fig_caption: yes
    toc: yes
vignette: |
  %\VignetteIndexEntry{motifbreakR: an Introduction} %\VignetteEngine{knitr::rmarkdown} \usepackage[utf8]{inputenc} \usepackage{graphicx}
---

<!-- see http://rmarkdown.rstudio.com/ for details in formatting -->
```{r style, echo = FALSE, results='hide', message=FALSE, warning=FALSE}
BiocStyle::markdown()
#BiocStyle::latex()
```
### Introduction
This document offers an introduction and overview of `r Githubpkg("Simon-Coetzee/motifBreakR")`, which allows the biologist to judge whether the sequence surrounding a polymorphism or mutation is a good match to known transcription factor binding sites, and how much information is gained or lost in one allele of the polymorphism relative to another or mutation vs. wildtype. `r Githubpkg("Simon-Coetzee/motifBreakR")` is flexible, giving a choice of algorithms for interrogation of genomes with motifs from public sources that users can choose from; these are 1) a weighted-sum, 2) log-probabilities, and 3) relative entropy. `r Githubpkg("Simon-Coetzee/motifBreakR")` can predict effects for novel or previously described variants in public databases, making it suitable for tasks beyond the scope of its original design. Lastly, it can be used to interrogate any genome curated within Bioconductor.

`r Githubpkg("Simon-Coetzee/motifBreakR")` works with position probability matrices (PPM). PPM are derived as the fractional occurrence of nucleotides A,C,G, and T at each position of a position frequency matrix (PFM). PFM are simply the tally of each nucleotide at each position across a set of aligned sequences. With a PPM, one can generate probabilities based on the genome, or more practically, create any number of position specific scoring matrices (PSSM) based on the principle that the PPM contains information about the likelihood of observing a particular nucleotide at a particular position of a true transcription factor binding site.

This guide includes a brief overview of the [processing flow](#flow), an example focusing more in depth on the [practical](#practical) aspect of using `r Githubpkg("Simon-Coetzee/motifBreakR")`, and finally a detailed section on the [scoring methods](#methods) employed by the package.


### Processing overview {#flow}
`r Githubpkg("Simon-Coetzee/motifBreakR")` may be used to interrogate SNPs or SNVs for their potential effect on transcription factor binding by examining how the two alleles of the variant effect the binding score of a motif.  The basic process is outlined in the figure below.

#### Outline of process

<!-- this is here in order to import a figure and to add a caption -->
<div class="figure" style="text-align: center">
  <img width="50%" src="motifbreakr-manuscript.png" />
  <p class="caption"><em>motifbreakR workflow: How inputs (trapezoids) are generated from R functions (rectangles). Diamonds represent decisions of the user.</em></p>
</div>
This straightforward process allows the interrogation of SNPs and SNVs in the context of the different species represented by `r Biocpkg("BSgenome")` packages (at least 22 different species) and allows the use of the full `r Biocpkg("MotifDb")` dataset that includes over 4200 motifs across 8 studies and 22 organisms that we have supplemented with over 2800 additional motifs across four additional studies in Humans see `data(encodemotif)`, `data(factorbook)`, `data(hocomoco)` and `data(homer)` for the additonal studies that we have included.

We can see here the full selection from MotifDb
```{r, message=FALSE, cache=TRUE, eval=FALSE}
library(MotifDb)
MotifDb
table(mcols(MotifDb)$organism, mcols(MotifDb)$dataSource)
```
```{r, message=FALSE, cache=TRUE, echo=FALSE}
library(MotifDb)
MotifDb
knitr::kable(table(mcols(MotifDb)$organism, mcols(MotifDb)$dataSource), format = "html", table.attr="class=\"table table-striped table-hover\"")
```

Here is the additional set of motifs that we have gathered and included in this package:
```{r, message=FALSE, cache=TRUE}
library(motifbreakR)
data(motifbreakR_motif)
motifbreakR_motif
```

### Methods {#methods}
`r Githubpkg("Simon-Coetzee/motifBreakR")` works with position probability matrices (PPM). PPM
are derived as the fractional occurrence of nucleotides A,C,G, and T at
each position of a position frequency matrix (PFM). PFM are simply the
tally of each nucleotide at each position across a set of aligned
sequences. With a PPM, one can generate probabilities based on the
genome, or more practically, create any number of position specific
scoring matrices (PSSM) based on the principle that the PPM contains
information about the likelihood of observing a particular nucleotide at
a particular position of a true transcription factor binding site. What
follows is a discussion of the three different algorithms that may be
employed in calls to the `r Githubpkg("Simon-Coetzee/motifBreakR")` function via the `method`
argument.

Suppose we have a frequency matrix $M$ of width $n$ (*i.e.* a
PPM as described above). Furthermore, we have a sequence $s$ also of
length $n$, such that
$s_{i} \in \{ A,T,C,G \}, i = 1,\ldots n$. Each column of
$M$ contains the frequencies of each letter in each position.

Commonly in the literature sequences are scored as the sum of log
probabilities:

#### Equation 1 {#eqn1}
$$F( s,M ) = \sum_{i = 1}^{n}{\log( \frac{M_{s_{i},i}}{b_{s_{i}}} )}$$

where $b_{s_{i}}$ is the background frequency of letter $s_{i}$ in
the genome of interest. This method can be specified by the user as
`method='log'`.

As an alternative to this method, we introduced a scoring method to
directly weight the score by the importance of the position within the
match sequence. This method of weighting is accessed by specifying
`method='IC'` (information content). A general representation
of this scoring method is given by:

#### Equation 2
$$F( s,M ) = p( s ) \cdot \omega_{M}$$

where $p_{s}$ is the scoring vector derived from sequence $s$ and matrix
$M$, and $w_{M}$ is a weight vector derived from $M$. First, we
compute the scoring vector of position scores $p$:

#### Equation 3
$$p( s ) = ( M_{s_{i},i} ) \textrm{   where   } \frac{i = 1,\ldots n}{s_{i} \in \{ A,C,G,T \}}$$

and second, for each $M$ a constant vector of weights
$\omega_{M} = ( \omega_{1},\omega_{2},\ldots,\omega_{n} )$.

There are two methods for producing $\omega_{M}$. The first, which we
call weighted sum, is the difference in the probabilities for the two
letters of the polymorphism (or variant), *i.e.*
$\Delta p_{s_{i}}$, or the difference of the maximum and minimum
values for each column of $M$:

#### Equation 4.1 {#eqn4.1}
$$\omega_{i} = \max \{ M_{i} \} - \min \{ M_{i} \}\textrm{    where      }i = 1,\ldots n$$

The second variation of this theme is to weight by relative entropy.
Thus the relative entropy weight for each column $i$ of the matrix is
given by:

#### Equation 4.2 {#eqn4.2}
$$\omega_{i} = \sum_{j \in \{ A,C,G,T \}}^{}{M_{j,i}\log( \frac{M_{j,i}}{b_{i}} )}\textrm{     where     }i = 1,\ldots n$$

where $b_{i}$ is again the background frequency of the letter $i$.

Thus, there are 3 possible algorithms to apply via the `method`
argument. The first is the standard summation of log probabilities
(`method='log'`). The second and third are the weighted sum and
information content methods (`method=default` and `method='IC'`) specified by
equations [4.1](#eqn4.1) and [4.2](#eqn4.2), respectively. `r Githubpkg("Simon-Coetzee/motifBreakR")` assumes a
uniform background nucleotide distribution ($b$) in equations [1](#eqn1) and
[4.2](#eqn4.1) unless otherwise specified by the user. Since we are primarily
interested in the difference between alleles, background frequency is
not a major factor, although it can change the results. Additionally,
inclusion of background frequency introduces potential bias when
collections of motifs are employed, since motifs are themselves
unbalanced with respect to nucleotide composition. With these cautions
in mind, users may override the uniform distribution if so desired. For
all three methods, `r Githubpkg("Simon-Coetzee/motifBreakR")` scores and reports the reference
and alternate alleles of the sequence
($F( s_{\textrm{REF}},M )$ and
$F( s_{\textrm{ALT}},M )$), and provides the matrix scores
$p_{s_{\textrm{REF}}}$ and $p_{s_{\textrm{ALT}}}$ of the SNP (or
variant). The scores are scaled as a fraction of scoring range 0-1 of
the motif matrix, $M$. If either of
$F( s_{\textrm{REF}},M )$ and
$F( s_{\textrm{ALT}},M )$ is greater than a user-specified
threshold (default value of 0.85) the SNP is reported. By default
`r Githubpkg("Simon-Coetzee/motifBreakR")` displays only strong effects
($\Delta p_{i} > 0.7$) but this behaviour can be
overridden.



## Vignette Info

Note the various macros within the `vignette` setion of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 
<!-- to use with html use fig.retina=2 for proper res figures -->
```{r test, fig.show='hold', fig.cap="test: this is a figure cap.", fig.retina=2} 
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
