---
title: 'motifbreakR: an Introduction'
author: "Simon G. Coetzee, Gerhard A. Coetzee and Dennis J. Hazelett"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: true
    highlight: textmate
    keep_md: yes
    theme: flatly
    toc: yes
    fig_width: 4.5
    fig_height: 4
vignette: |
  %\VignetteIndexEntry{motifbreakR: an Introduction} %\VignetteEngine{knitr::rmarkdown}
---

<!-- see http://rmarkdown.rstudio.com/ for details in formatting -->
```{r style, echo = FALSE, results='hide', message=FALSE, warning=FALSE}
BiocStyle::markdown()
options(width=100)
```
### Introduction
This document offers an introduction and overview of `r Githubpkg("Simon-Coetzee/motifBreakR")`, which allows the biologist to judge whether the sequence surrounding a polymorphism or mutation is a good match to known transcription factor binding sites, and how much information is gained or lost in one allele of the polymorphism relative to another or mutation vs. wildtype. `r Githubpkg("Simon-Coetzee/motifBreakR")` is flexible, giving a choice of algorithms for interrogation of genomes with motifs from public sources that users can choose from; these are 1) a weighted-sum, 2) log-probabilities, and 3) relative entropy. `r Githubpkg("Simon-Coetzee/motifBreakR")` can predict effects for novel or previously described variants in public databases, making it suitable for tasks beyond the scope of its original design. Lastly, it can be used to interrogate any genome curated within Bioconductor.

`r Githubpkg("Simon-Coetzee/motifBreakR")` works with position probability matrices (PPM). PPM are derived as the fractional occurrence of nucleotides A,C,G, and T at each position of a position frequency matrix (PFM). PFM are simply the tally of each nucleotide at each position across a set of aligned sequences. With a PPM, one can generate probabilities based on the genome, or more practically, create any number of position specific scoring matrices (PSSM) based on the principle that the PPM contains information about the likelihood of observing a particular nucleotide at a particular position of a true transcription factor binding site.

This guide includes a brief overview of the [processing flow](#flow), an example focusing more in depth on the [practical](#practical) aspect of using `r Githubpkg("Simon-Coetzee/motifBreakR")`, and finally a detailed section on the [scoring methods](#methods) employed by the package.


### Processing overview {#flow}
`r Githubpkg("Simon-Coetzee/motifBreakR")` may be used to interrogate SNPs or SNVs for their potential effect on transcription factor binding by examining how the two alleles of the variant effect the binding score of a motif.  The basic process is outlined in the figure below.

#### Outline of process

<!-- this is here in order to import a figure and to add a caption -->
<div class="figure" style="text-align: center">
  <img width="50%" src="motifbreakr-manuscript.png" />
  <p class="caption"><em>motifbreakR workflow: How inputs (trapezoids) are generated from R functions (rectangles). Diamonds represent decisions of the user.</em></p>
</div>
This straightforward process allows the interrogation of SNPs and SNVs in the context of the different species represented by `r Biocpkg("BSgenome")` packages (at least 22 different species) and allows the use of the full `r Biocpkg("MotifDb")` data set that includes over 4200 motifs across 8 studies and 22 organisms that we have supplemented with over 2800 additional motifs across four additional studies in Humans see `data(encodemotif)`[^encodemotif], `data(factorbook)`[^factorbook], `data(hocomoco)`[^hocomoco] and `data(homer)`[^homer] for the additional studies that we have included.

[^encodemotif]: Website: [encode-motifs](http://compbio.mit.edu/encode-motifs/) Paper: [doi:10.1093/nar/gkt1249](http://dx.doi.org/10.1093/nar/gkt1249)
[^factorbook]: Website: [Factorbook](http://factorbook.org/mediawiki/index.php/Welcome_to_factorbook) Paper: [doi:10.1101/gr.139105.112](http://dx.doi.org/10.1101/gr.139105.112)
[^hocomoco]: Website: [HOCOMOCO](http://autosome.ru/HOCOMOCO/) Paper: [doi:10.1093/nar/gks1089](http://dx.doi.org/10.1093/nar/gks1089)
[^homer]: Website: [Homer](http://homer.salk.edu/homer/index.html) Paper: [doi:10.1016/j.molcel.2010.05.004](http://dx.doi.org/10.1016/j.molcel.2010.05.004)

Practically `r Githubpkg("Simon-Coetzee/motifBreakR")` has involves three phases.

1. *Read in Single Nucleotide Variants:* The first step is to read in the list of variants.  Variants can be either a list of rs#'s if your SNPs are represented in one of the SNPlocs packages, or may be included as a [.bed](http://www.genome.ucsc.edu/FAQ/FAQformat.html#format1) file with a specifically formatted name field. We then transform this input such that it may be read my `r Githubpkg("Simon-Coetzee/motifBreakR")`.
2. *Find Broken Motifs:* Next we present `r Githubpkg("Simon-Coetzee/motifBreakR")` with the input generated in the previous step, along with a set of motifs formatted as class `MotifList`, and your preferred [scoring method](#methods). 
3. *Visualize SNPs and motifs:* Finally we can visualize which motifs are broken by any individual SNP using our plotting function.

### How To Use motifbreakR: A practical Example {#practical}
#### *Step 1* | Read in Single Nucleotide Variants
Variants can be input either as a list of rs#'s or as a [.bed](http://www.genome.ucsc.edu/FAQ/FAQformat.html#format1). The main factor determining which you will use is if your variants have rs#'s that are included in one of the Bioconductor SNPlocs packages.  The present selection is seen here:
```{r, message=FALSE, cache=TRUE}
library(BSgenome)
available.SNPs()
```
For cases where your rs#'s are not available in a SNPlocs package, or you have novel variants that are not cataloged at all, variants may be entered in [BED](http://www.genome.ucsc.edu/FAQ/FAQformat.html#format1) format as seen below:
```{r, message=FALSE, cache=TRUE}
snps.file <- system.file("extdata", "snps.bed", package = "motifbreakR")
read.delim(snps.file, header = FALSE)
```
Our requirements for the BED file are that it must include `chromosome`, `start`, `end`, `name`, `score` and `strand` fields -- where the name field is required to be in one of two formats, either an rs# that is present in a SNPlocs package, or in the form `chromosome:position:referenceAllele:alternateAllele` e.g., `chr2:12594018:G:A`.  It is also essential that the fields are TAB seperated, not a mixture of tabs and spaces.
More to the point here are the two methods for reading in the variants.

##### SNPs from RS#:
```{r, message=FALSE, cache=TRUE}
library(SNPlocs.Hsapiens.dbSNP.20120608) # dbSNP137 in hg19
library(BSgenome.Hsapiens.UCSC.hg19)     # hg19 genome
snps.file <- system.file("extdata", "pca.enhancer.snps", package = "motifbreakR")
snps <- as.character(read.table(snps.file)[,1])
head(snps)
snps.mb <- snps.from.rsid(snps,
                          dbSNP = SNPlocs.Hsapiens.dbSNP.20120608,
                          search.genome = BSgenome.Hsapiens.UCSC.hg19)
snps.mb
```
A far greater variety of variants may be read into `r Githubpkg("Simon-Coetzee/motifBreakR")` via the `snps.from.bed` function. In fact `r Githubpkg("Simon-Coetzee/motifBreakR")` will work with any organism present as a Bioconductor `r Biocpkg("BSgenome")` package. This includes 76 genomes representing 22 species.
```{r, message=FALSE, cache=TRUE}
library(BSgenome)
genomes <- available.genomes()
length(genomes)
genomes
```

##### SNPs from BED formatted file:
```{r, message=TRUE, cache=TRUE}
snps.bed.file <- system.file("extdata", "snps.bed", package = "motifbreakR")
# see the contents
read.table(snps.bed.file, header = FALSE)
```
Seeing as we have some snps listed by thier rs#'s we can query those by including a SNPlocs object as an argument to `snps.from.bed`
```{r, message=TRUE, cache=TRUE}
#import the BED file
snps.mb.frombed <- snps.from.bed(snps.bed.file,
                                 dbSNP = SNPlocs.Hsapiens.dbSNP.20120608,
                                 search.genome = BSgenome.Hsapiens.UCSC.hg19)
snps.mb.frombed
```
We see also that one of our custom variants `chr2:12594018:G:A` was actually already included in dbSNP, and was therefor annotated in the output as `rs7601289`

If our BED file includes no rs#'s, then we may omit the `dbSNP` arguement from the function
```{r, message=FALSE, cache=TRUE}
library(BSgenome.Drerio.UCSC.danRer7)
snps.bedfile.nors <- system.file("extdata", "danRer.bed", package = "motifbreakR")
read.table(snps.bedfile.nors, header = FALSE)
snps.mb.frombed <- snps.from.bed(snps.bedfile.nors,
                                 search.genome = BSgenome.Drerio.UCSC.danRer7)
snps.mb.frombed
```

#### *Step 2* | Find Broken Motifs
Now that we have our data in the required format, we may continue to the task at hand, and determine which variants modify potential transcription factor binding. An important element of this task is identifying a set of transcription factor binding motifs that we wish to query. Fortunately `r Biocpkg("MotifDb")` includes a large selection of motifs across multiple species that we can see here:
```{r, message=FALSE, cache=TRUE}
library(MotifDb)
MotifDb
```
```{r, message=FALSE, cache=TRUE, eval=FALSE}
### Here we can see which organisms are availible under which sources
### in MotifDb
table(mcols(MotifDb)$organism, mcols(MotifDb)$dataSource)
```
```{r, message=FALSE, cache=TRUE, echo=FALSE}
knitr::kable(table(mcols(MotifDb)$organism, mcols(MotifDb)$dataSource), format = "html", table.attr="class=\"table table-striped table-hover\"")
```

And here we can see the additional set of motifs that have been gathered and included in this package:
```{r, message=FALSE, cache=TRUE}
library(motifbreakR)
data(motifbreakR_motif)
motifbreakR_motif
```
The different studies included in this data set may be called individually; for example:
```{r, message=FALSE, cache=TRUE}
data(hocomoco)
hocomoco
```
See `?motifbreakR_motif` for more information and citations.
We can use


### Methods {#methods}
`r Githubpkg("Simon-Coetzee/motifBreakR")` works with position probability matrices (PPM). PPM
are derived as the fractional occurrence of nucleotides A,C,G, and T at
each position of a position frequency matrix (PFM). PFM are simply the
tally of each nucleotide at each position across a set of aligned
sequences. With a PPM, one can generate probabilities based on the
genome, or more practically, create any number of position specific
scoring matrices (PSSM) based on the principle that the PPM contains
information about the likelihood of observing a particular nucleotide at
a particular position of a true transcription factor binding site. What
follows is a discussion of the three different algorithms that may be
employed in calls to the `r Githubpkg("Simon-Coetzee/motifBreakR")` function via the `method`
argument.

Suppose we have a frequency matrix $M$ of width $n$ (*i.e.* a
PPM as described above). Furthermore, we have a sequence $s$ also of
length $n$, such that
$s_{i} \in \{ A,T,C,G \}, i = 1,\ldots n$. Each column of
$M$ contains the frequencies of each letter in each position.

Commonly in the literature sequences are scored as the sum of log
probabilities:

#### Equation 1 {#eqn1}
$$F( s,M ) = \sum_{i = 1}^{n}{\log( \frac{M_{s_{i},i}}{b_{s_{i}}} )}$$

where $b_{s_{i}}$ is the background frequency of letter $s_{i}$ in
the genome of interest. This method can be specified by the user as
`method='log'`.

As an alternative to this method, we introduced a scoring method to
directly weight the score by the importance of the position within the
match sequence. This method of weighting is accessed by specifying
`method='IC'` (information content). A general representation
of this scoring method is given by:

#### Equation 2
$$F( s,M ) = p( s ) \cdot \omega_{M}$$

where $p_{s}$ is the scoring vector derived from sequence $s$ and matrix
$M$, and $w_{M}$ is a weight vector derived from $M$. First, we
compute the scoring vector of position scores $p$:

#### Equation 3
$$p( s ) = ( M_{s_{i},i} ) \textrm{   where   } \frac{i = 1,\ldots n}{s_{i} \in \{ A,C,G,T \}}$$

and second, for each $M$ a constant vector of weights
$\omega_{M} = ( \omega_{1},\omega_{2},\ldots,\omega_{n} )$.

There are two methods for producing $\omega_{M}$. The first, which we
call weighted sum, is the difference in the probabilities for the two
letters of the polymorphism (or variant), *i.e.*
$\Delta p_{s_{i}}$, or the difference of the maximum and minimum
values for each column of $M$:

#### Equation 4.1 {#eqn4.1}
$$\omega_{i} = \max \{ M_{i} \} - \min \{ M_{i} \}\textrm{    where      }i = 1,\ldots n$$

The second variation of this theme is to weight by relative entropy.
Thus the relative entropy weight for each column $i$ of the matrix is
given by:

#### Equation 4.2 {#eqn4.2}
$$\omega_{i} = \sum_{j \in \{ A,C,G,T \}}^{}{M_{j,i}\log( \frac{M_{j,i}}{b_{i}} )}\textrm{     where     }i = 1,\ldots n$$

where $b_{i}$ is again the background frequency of the letter $i$.

Thus, there are 3 possible algorithms to apply via the `method`
argument. The first is the standard summation of log probabilities
(`method='log'`). The second and third are the weighted sum and
information content methods (`method=default` and `method='IC'`) specified by
equations [4.1](#eqn4.1) and [4.2](#eqn4.2), respectively. `r Githubpkg("Simon-Coetzee/motifBreakR")` assumes a
uniform background nucleotide distribution ($b$) in equations [1](#eqn1) and
[4.2](#eqn4.1) unless otherwise specified by the user. Since we are primarily
interested in the difference between alleles, background frequency is
not a major factor, although it can change the results. Additionally,
inclusion of background frequency introduces potential bias when
collections of motifs are employed, since motifs are themselves
unbalanced with respect to nucleotide composition. With these cautions
in mind, users may override the uniform distribution if so desired. For
all three methods, `r Githubpkg("Simon-Coetzee/motifBreakR")` scores and reports the reference
and alternate alleles of the sequence
($F( s_{\textrm{REF}},M )$ and
$F( s_{\textrm{ALT}},M )$), and provides the matrix scores
$p_{s_{\textrm{REF}}}$ and $p_{s_{\textrm{ALT}}}$ of the SNP (or
variant). The scores are scaled as a fraction of scoring range 0-1 of
the motif matrix, $M$. If either of
$F( s_{\textrm{REF}},M )$ and
$F( s_{\textrm{ALT}},M )$ is greater than a user-specified
threshold (default value of 0.85) the SNP is reported. By default
`r Githubpkg("Simon-Coetzee/motifBreakR")` displays only strong effects
($\Delta p_{i} > 0.7$) but this behavior can be
overridden.
